service/storage_service.cc

```diff
@@ -232,62 +256,20 @@ void storage_service::prepare_to_join(std::vector<inet_address> loaded_endpoints
         if (db::system_keyspace::bootstrap_complete()) {
             throw std::runtime_error("Cannot replace address with a node that is already bootstrapped");
         }
-        if (!is_auto_bootstrap()) {
-            throw std::runtime_error("Trying to replace_address with auto_bootstrap disabled will not work, check your configuration");
-        }
-
-        std::tie(_bootstrap_tokens, _cdc_streams_ts) = prepare_replacement_info(loaded_peer_features).get0();
+        std::tie(_bootstrap_tokens, _cdc_streams_ts) = prepare_replacement_info(initial_contact_nodes, loaded_peer_features).get0();
         auto replace_address = db().local().get_replace_address();
         replacing_a_node_with_same_ip = replace_address && *replace_address == get_broadcast_address();
         replacing_a_node_with_diff_ip = replace_address && *replace_address != get_broadcast_address();
     } else if (should_bootstrap()) {
-        check_for_endpoint_collision(loaded_peer_features).get();
+        check_for_endpoint_collision(initial_contact_nodes, loaded_peer_features).get();
     } else {
-        auto seeds = _gossiper.get_seeds();
-        auto my_ep = get_broadcast_address();
         auto local_features = _feature_service.known_feature_set();
-
-        if (seeds.count(my_ep)) {
-            // This node is a seed node
-            if (loaded_peer_features.empty()) {
-                // This is a competely new seed node, skip the check
-                slogger.info("Checking remote features skipped, since this node is a new seed node which knows nothing about the cluster");
-            } else {
-                // This is a existing seed node
-                if (seeds.size() == 1) {
-                    // This node is the only seed node, check features with system table
-                    slogger.info("Checking remote features with system table, since this node is the only seed node");
-                    _gossiper.check_knows_remote_features(local_features, loaded_peer_features);
-                } else {
-                    // More than one seed node in the seed list, do shadow round with other seed nodes
-                    try {
-                        slogger.info("Checking remote features with gossip and system tables");
-                        _gossiper.do_shadow_round().get();
-                    } catch (...) {
-                        slogger.info("Shadow round failed with {}, checking remote features with system tables only",
-                                std::current_exception());
-                        _gossiper.finish_shadow_round();
-                    }
-
-                    _gossiper.check_knows_remote_features(local_features, loaded_peer_features);
-                    _gossiper.reset_endpoint_state_map().get();
-                    for (auto ep : loaded_endpoints) {
-                        _gossiper.add_saved_endpoint(ep);
-                    }
-                }
-            }
```

Comment:
How does this removal work in a mixed environment before all nodes are upgraded?

Reply:
For a mixed cluster, the shadow round will fallback to the old method.  It is a upgrade operation, so this is going to be an exiting cluster. The check should just work.